### ISSUE TITLE
Website integration issues

### ISSUE THREAD

**Issue by d4708** [zeronet-conservancy/zeronet-conservancy#45]
I'm trying to create 0net mirror of https://twisterarmy.github.io website.

But I have some problems with the rendering:

1. Getting 404 error for css, fonts and images addresses, because website running in the frame and base tag could not be applied
2. All URL links configured to use extension-less URI (without .html postfix) so can I apply some solution, without changing in the website code

The address of site is 1PdfH9GQGT5ahR9WwYMzZCyBNdxA2pFdg7
Maybe Ctrl+Shift+i say more.

Thanks for any ideas!

**Comment by caryoscelus** [comment]
i can see your site , but yeah a lot of images etc seem missing . you can try removing leading '/' for all core urls , that should solve at least main page issue . otherwise i'll have to remember proper technique for that kind of things

also , your site is too big by 0net standards , most big files should be optional (i know , additional headache reading manuals)

> All URL links configured to use extension-less URI (without .html postfix) so can I apply some solution, without changing in the website code

are they leading to directories with `index.html` or smth else?

**Comment by d4708** [comment]
> removing leading '/' for all core urls
yep, I just fixed that (some temporarily solution from previous story)

> are they leading to directories with index.html or smth else?
nope, it is just a files with html extensions. that is [github hosted website](https://github.com/twisterarmy/twisterarmy.github.io), github allows extension-less requests, so I thinking how to make it accessible trough zero interface, seems I need to return extensions or make it as directories, but so won't...

anyway, another issue - when I'm trying to open extension-less link, files just downloading in FF, maybe some headers missed, FF thinks that file for download..

also I can't download files like this one:
/1PdfH9GQGT5ahR9WwYMzZCyBNdxA2pFdg7/download/twister-ta-1.1-amd64.deb

with error alert
`Connection with UiServer Websocket was lost. Reconnecting...`

> your site is too big by 0net standards

thank you, did't knew that. anyway strange, because I wish to create some file collection for sharing (in another website)

**Comment by caryoscelus** [comment]
> anyway, another issue - when I'm trying to open extension-less link, files just downloading in FF, maybe some headers missed, FF thinks that file for download..

that (downloading part) can be tweaked in 0net code , i've seen related code recently .. but it's surprising that github adds .html extension file requests (i don't know , is that a part of jekyll spec?) and i'm not sure what would be the best solution

> thank you, did't knew that. anyway strange, because I wish to create some file collection for sharing (in another website)

well , that's in order to make sure more users seed the main site at least (at least back in the days , a lot of people would download random sites and just keep them seeding without having any interest in them) . optional files can be as big as they need be , but only people who need them (or wanting to support your site) will download them

> also I can't download files like this one:
> /1PdfH9GQGT5ahR9WwYMzZCyBNdxA2pFdg7/download/twister-ta-1.1-amd64.deb

http://127.0.0.1:43110/list/1PdfH9GQGT5ahR9WwYMzZCyBNdxA2pFdg7/ shows that there's no download directory (or it hasn't been downloaded, maybe `content.json` issues)

**Comment by caryoscelus** [comment]
> http://127.0.0.1:43110/list/1PdfH9GQGT5ahR9WwYMzZCyBNdxA2pFdg7/ shows that there's no download directory (or it hasn't been downloaded, maybe content.json issues)

> twister-ta-1.1-amd64.deb
> twister-ta-1.1-armhf.deb

files are just missing on my sync yet , so disregard that

**Comment by d4708** [comment]
Have updated URI's by removing root slash from the addresses , that allows to generate links that starts with website address (1PdfH9GQGT5ahR9WwYMzZCyBNdxA2pFdg7) prefix

About html extensions, I won't to place the pages with index files of it folders, because browsers attach the slash as postfix of address on the web production. And I have some documentation page there, where it sections linked using anchors and I won't to see the URL's like page/#anchor. 

That's also not pretty correct for SEO reasons.

One more issue I'm getting is the CORS errors for the fonts including:

```
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource
...
Downloadable font: download failed (font-family: "bootstrap-icons" style:normal weight:400 stretch:100 src index:0): bad URI or cross-site access not allowed source: http://127.0.0.1:43110/1PdfH9GQGT5ahR9WwYMzZCyBNdxA2pFdg7/css/font/bi/set.woff2
```

Maybe as the solution I need to maintain additional branch just for zeronet iframe...

**Comment by caryoscelus** [comment]
> One more issue I'm getting is the CORS errors for the fonts including:

let me check how i managed that on my site

**Comment by caryoscelus** [comment]
> Downloadable font: download failed (font-family: "bootstrap-icons" style:normal weight:400 stretch:100 src index:0): bad URI or cross-site access not allowed source: http://127.0.0.1:43110/1PdfH9GQGT5ahR9WwYMzZCyBNdxA2pFdg7/css/font/bi/set.woff2

heh , it cannot recognize that it's the same website ;)

anyway , what i did is using relative urls like `../../fonts/<font-name>`

**Comment by d4708** [comment]
So, after more than ten years of the web experience, had spent about few days to make simple html site compatible with 0net framing.

Bellow I'll describe just my own solution, because this way looks for me simpler in the administration context.
In my case, I see pretty usable automation script, that insert Website ID prefixes to the document root and add .html postfixes to the end of URIs.

For these reasons, I have unified all the links to the format, where they are starting from the root directory of website. The same solution for the css paths. Just an example bellow:
https://github.com/twisterarmy/twisterarmy.github.io/commit/ac9444127e03287dc442022c3f79e14e024bba2f

Maybe another solutions available, but in my case, I have multilingual website, where each the language placed into the sub-directory, and I can't define relative paths commonly, or if I do that, some of the links, like ../uk not works from the internal pages like /uk/page as it relates homepage with the localhost :43110 and the Website ID skipping by iframe (just link to the 0net homepage, not website page)

As result, beside the github pages repository, I have started to maintain separated zeronet branch, it's no ways for me to keep cross-compatible single version.

--

Some another subjects:

Issue 1
Not clickable external links like this one in the top website menu 
![2022-05-14_20-04](https://user-images.githubusercontent.com/96319366/168441611-b2919b06-a1bb-4bb9-835a-6f577b3be408.png)

Issue 2
File downloading links not works (even locally)
http://127.0.0.1/1PdfH9GQGT5ahR9WwYMzZCyBNdxA2pFdg7/download/twister-ta-1.1-amd64.deb

p.s I have just published updates, so available for connections now.

**Comment by caryoscelus** [comment]
> So, after more than ten years of the web experience, had spent about few days to make simple html site compatible with 0net framing.

i guess that answers a lot of questions as to why 0net has been stagnating :/ and needs heavy rewriting

sites looks visually cool now

i'll read & answer on the details tomorrow

**Comment by caryoscelus** [comment]
just add something like "optional": ".*deb" to the main `content.json` . it's regexp , so you can write ".*\.deb" if you want to be slightly more precise

**Comment by d4708** [comment]
like this one, beside the top-level keys? I just supposed it should be json key maybe inside the file specified construction..

thank you!

```
"optional": ".*\.deb",
 "ignore": "",
 "inner_path": "content.json",
```

**Comment by d4708** [comment]
added
```
 "optional": "(download/.*)",
```
after sign it is new constructions generated, looks that done.

```
 "files_optional": {
  "download/twister-ta-1.1-amd64.deb": {
...
```

p.s. about piecemap.msgpack ones
https://github.com/HelloZeroNet/ZeroNet/issues/1485

**Comment by d4708** [comment]
https://github.com/zeronet-conservancy/zeronet-conservancy/issues/63#issuecomment-1132220255
